python/instrumentation/openinference-instrumentation-llama-index/src/openinference/instrumentation/llama_index/_handler.py

```diff
@@ -0,0 +1,507 @@
+    ReRankEndEvent,
+    ReRankStartEvent,
+)
+from llama_index.core.instrumentation.events.retrieval import (
+    RetrievalEndEvent,
+    RetrievalStartEvent,
+)
+from llama_index.core.instrumentation.events.span import (
+    SpanDropEvent,
+)
+from llama_index.core.instrumentation.events.synthesis import (
+    GetResponseEndEvent,
+    GetResponseStartEvent,
+    SynthesizeEndEvent,
+    SynthesizeStartEvent,
+)
+from llama_index.core.instrumentation.span import BaseSpan
+from llama_index.core.instrumentation.span_handlers import BaseSpanHandler
+from llama_index.core.schema import QueryType
+
+logger = logging.getLogger(__name__)
+
+
+class _Span(
+    BaseSpan,
+    extra=Extra.allow,
+    keep_untouched=(singledispatchmethod, property),
+):
+    _otel_span: trace_api.Span = PrivateAttr()
+    _span_kind: Optional[str] = PrivateAttr()
+    _first_token_timestamp: Optional[datetime] = PrivateAttr()
+
+    def __init__(
+        self,
+        otel_span: trace_api.Span,
+        span_kind: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._otel_span = otel_span
+        self._span_kind = span_kind
+        self._first_token_timestamp = None
+
+    def __setitem__(self, key: str, value: AttributeValue) -> None:
+        self._otel_span.set_attribute(key, value)
+
+    def record_exception(self, exception: BaseException) -> None:
+        self._otel_span.record_exception(exception)
+
+    def end(self, status: trace_api.Status) -> None:
+        self[OPENINFERENCE_SPAN_KIND] = self._span_kind or CHAIN
+        self._otel_span.set_status(status=status)
+        self._otel_span.end()
+
+    @property
+    def context(self) -> context_api.Context:
+        return trace_api.set_span_in_context(self._otel_span)
+
+    @singledispatchmethod
+    def process_event(self, event: BaseEvent) -> None:
+        logger.warning(f"Unhandled event of type {event.__class__.__qualname__}")
+
+    @process_event.register
+    def _(self, event: AgentChatWithStepStartEvent) -> None:
+        if not self._span_kind:
+            self._span_kind = AGENT
+        self[INPUT_VALUE] = event.user_msg
+
+    @process_event.register
+    def _(self, event: AgentChatWithStepEndEvent) -> None:
+        self[OUTPUT_VALUE] = str(event.response)
+
+    @process_event.register
+    def _(self, event: AgentRunStepStartEvent) -> None:
+        if not self._span_kind:
+            self._span_kind = AGENT
+        if input := event.input:
+            self[INPUT_VALUE] = input
+
+    @process_event.register
+    def _(self, event: AgentRunStepEndEvent) -> None:
+        # FIXME: not sure what to do here
+        ...
+
+    @process_event.register
+    def _(self, event: AgentToolCallEvent) -> None:
+        tool = event.tool
+        if name := tool.name:
+            self[TOOL_NAME] = name
+        self[TOOL_DESCRIPTION] = tool.description
+        self[TOOL_PARAMETERS] = json.dumps(tool.get_parameters_dict())
```

Comment:
are there more params we could capture?

Reply:
Yes, but they are on the instance and we would need to figure out how to find and exclude the API_KEYs so we are not grabbing too much.